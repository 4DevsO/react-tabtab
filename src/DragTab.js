import React from 'react';
import {SortableElement} from 'react-sortable-hoc';
import Tab from './Tab';

const DragTabElement = SortableElement(({children, ...props}) => {
  return (
    <Tab index={props.tabIndex} {...props}>
      {children}
    </Tab>
  )
});

class DragTab extends React.Component {
  render() {
    const {children, ...props} = this.props;
    return (
      <DragTabElement {...props}>
        {children}
      </DragTabElement>
    )
  }
}

export default DragTab;