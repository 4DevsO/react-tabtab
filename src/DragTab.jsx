import React from 'react';
import ReactDOM from 'react-dom';
import { DragSource, DropTarget } from 'react-dnd';
import ItemTypes from './ItemTypes';
import Tab from './Tab';

const tabSource = {
  beginDrag: function(props) {
    props.beginDrag();
    return {
      tabKey: props.tabKey
    }
  }
}

// reference from https://github.com/gaearon/react-dnd/tree/master/examples/04%20Sortable/Simple
const tabTarget = {
  hover(props, monitor, component) {
    const dragIndex = monitor.getItem().tabKey;
    const hoverIndex = props.tabKey;

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = ReactDOM.findDOMNode(component).getBoundingClientRect();

    // Get horizontal middle
    const hoverMiddleX = (hoverBoundingRect.right - hoverBoundingRect.left) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    const hoverClientX = hoverBoundingRect.right - clientOffset.x;

    // Dragging left
    if (dragIndex < hoverIndex && hoverClientX < hoverMiddleX) {
      return;
    }

    // Dragging right
    if (dragIndex > hoverIndex && hoverClientX > hoverMiddleX) {
      return;
    }

    props.moveTab(dragIndex, hoverIndex);
    // change the dragging tab index
    monitor.getItem().tabKey = hoverIndex;
  }
};

@DropTarget(ItemTypes.TAB, tabTarget, connect => ({
  connectDropTarget: connect.dropTarget()
}))
@DragSource(ItemTypes.TAB, tabSource, (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  isDragging: monitor.isDragging()
}))

export default class DragTab extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {

    const {connectDragSource, connectDropTarget } = this.props;

    return connectDragSource(connectDropTarget(
      <span>
        <Tab {...this.props}/>
      </span>
    ));
  }
}