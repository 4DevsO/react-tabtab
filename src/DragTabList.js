import React from 'react';
import {SortableContainer} from 'react-sortable-hoc';
import TabList from './TabList';

const DragTabContainer = SortableContainer(({children, ...props}) => {
  return (
    <TabList {...props}>
      {children}
    </TabList>
  );
});

export default class DragTabList extends React.Component {
  constructor(props) {
    super(props);
    this.onSortEnd = this.onSortEnd.bind(this);
  }

  onSortEnd({oldIndex, newIndex}) {
    const {activeIndex, handleActiveIndex, handleTabSequence} = this.props;
    if (oldIndex === newIndex) {
      if (activeIndex !== oldIndex) {
        handleActiveIndex(oldIndex);
      }
    } else {
      console.log('switch: ', oldIndex, newIndex)
      handleTabSequence({oldIndex, newIndex})
    }
  }

  render() {
    const {children, ...props} = this.props; // eslint-disable-line
    return (
      <DragTabContainer onSortEnd={this.onSortEnd}
                        axis='x'
                        lockAxis='x'
                        {...props}>
        {children}
      </DragTabContainer>
    );
  }
}

DragTabList.displayName = 'DragTabList';
