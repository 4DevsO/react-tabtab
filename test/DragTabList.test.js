import DragTabList from '../src/DragTabList';
import DragTab from '../src/DragTab';
import tabListTest from './tabListTest';

describe('DragTabList', () => {
  tabListTest('DragTab', DragTabList, DragTab);
})
