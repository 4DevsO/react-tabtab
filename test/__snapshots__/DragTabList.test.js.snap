// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragTabList custom ListStyle style 1`] = `
.c0 {
  background-color: white;
  text-align: left;
  position: relative;
  white-space: nowrap;
  overflow: hidden;
  width: auto;
  border-bottom: 1px solid #eee;
  padding: 0 0px 0 0px;
}

.c1 {
  overflow: hidden;
}

.c2 {
  padding-left: 0;
  position: relative;
  margin: 0;
  list-style: none;
  display: inline-block;
  -webkit-transition: -webkit-transform .3s cubic-bezier(.42,0,.58,1);
  -webkit-transition: transform .3s cubic-bezier(.42,0,.58,1);
  transition: transform .3s cubic-bezier(.42,0,.58,1);
}

.c3 {
  display: inline-block;
  color: #007bff;
  border-top-left-radius: .25rem;
  border-top-right-radius: .25rem;
  border: 1px solid transparent;
  padding: 10px 15px;
  -webkit-transition: color .3s cubic-bezier(.645,.045,.355,1);
  transition: color .3s cubic-bezier(.645,.045,.355,1);
  -webkit-transition: background-color .3s cubic-bezier(.645,.045,.355,1);
  transition: background-color .3s cubic-bezier(.645,.045,.355,1);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c3:hover {
  cursor: pointer;
  color: black;
  border-color: #ddd #ddd #fff;
}

<DragTabList
  customStyle={
    Object {
      "CustomTabList": [Function],
    }
  }
>
  <sortableList
    axis="x"
    customStyle={
      Object {
        "CustomTabList": [Function],
      }
    }
    distance={2}
    getHelperDimensions={[Function]}
    hideSortableGhost={true}
    lockAxis="x"
    lockOffset="50%"
    lockToContainerEdges={false}
    onSortEnd={[Function]}
    pressDelay={0}
    pressThreshold={5}
    shouldCancelStart={[Function]}
    transitionDuration={300}
    useWindowAsScrollContainer={false}
  >
    <Component
      customStyle={
        Object {
          "CustomTabList": [Function],
        }
      }
    >
      <TabList
        customStyle={
          Object {
            "CustomTabList": [Function],
          }
        }
      >
        <div>
          <styled.div
            hasExtraButton={false}
          >
            <div
              className="c0"
            >
              <styled.div
                innerRef={[Function]}
              >
                <div
                  className="c1"
                >
                  <styled.div
                    innerRef={[Function]}
                  >
                    <div
                      className="c2"
                    >
                      <DragTab
                        active={false}
                        index={0}
                        key="0/.0"
                        tabIndex={0}
                      >
                        <sortableElement
                          active={false}
                          collection={0}
                          index={0}
                          tabIndex={0}
                        >
                          <Component
                            active={false}
                            tabIndex={0}
                          >
                            <Tab
                              active={false}
                              index={0}
                              tabIndex={0}
                            >
                              <styled.div
                                active={false}
                                innerRef={[Function]}
                                onClick={[Function]}
                              >
                                <div
                                  className="c3"
                                  onClick={[Function]}
                                >
                                  Tab1
                                </div>
                              </styled.div>
                            </Tab>
                          </Component>
                        </sortableElement>
                      </DragTab>
                    </div>
                  </styled.div>
                </div>
              </styled.div>
            </div>
          </styled.div>
        </div>
      </TabList>
    </Component>
  </sortableList>
</DragTabList>
`;

exports[`DragTabList render DragTab List 1`] = `
.c0 {
  background-color: white;
  text-align: left;
  position: relative;
  white-space: nowrap;
  overflow: hidden;
  width: auto;
  border-bottom: 1px solid #eee;
  padding: 0 0px 0 0px;
}

.c1 {
  overflow: hidden;
}

.c2 {
  padding-left: 0;
  position: relative;
  margin: 0;
  list-style: none;
  display: inline-block;
  -webkit-transition: -webkit-transform .3s cubic-bezier(.42,0,.58,1);
  -webkit-transition: transform .3s cubic-bezier(.42,0,.58,1);
  transition: transform .3s cubic-bezier(.42,0,.58,1);
}

.c3 {
  display: inline-block;
  color: #007bff;
  border-top-left-radius: .25rem;
  border-top-right-radius: .25rem;
  border: 1px solid transparent;
  padding: 10px 15px;
  -webkit-transition: color .3s cubic-bezier(.645,.045,.355,1);
  transition: color .3s cubic-bezier(.645,.045,.355,1);
  -webkit-transition: background-color .3s cubic-bezier(.645,.045,.355,1);
  transition: background-color .3s cubic-bezier(.645,.045,.355,1);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c3:hover {
  cursor: pointer;
  color: black;
  border-color: #ddd #ddd #fff;
}

<DragTabList
  customStyle={Object {}}
>
  <sortableList
    axis="x"
    customStyle={Object {}}
    distance={2}
    getHelperDimensions={[Function]}
    hideSortableGhost={true}
    lockAxis="x"
    lockOffset="50%"
    lockToContainerEdges={false}
    onSortEnd={[Function]}
    pressDelay={0}
    pressThreshold={5}
    shouldCancelStart={[Function]}
    transitionDuration={300}
    useWindowAsScrollContainer={false}
  >
    <Component
      customStyle={Object {}}
    >
      <TabList
        customStyle={Object {}}
      >
        <div>
          <styled.div
            hasExtraButton={false}
          >
            <div
              className="c0"
            >
              <styled.div
                innerRef={[Function]}
              >
                <div
                  className="c1"
                >
                  <styled.div
                    innerRef={[Function]}
                  >
                    <div
                      className="c2"
                    >
                      <DragTab
                        active={false}
                        index={0}
                        key="0/.0"
                        tabIndex={0}
                      >
                        <sortableElement
                          active={false}
                          collection={0}
                          index={0}
                          tabIndex={0}
                        >
                          <Component
                            active={false}
                            tabIndex={0}
                          >
                            <Tab
                              active={false}
                              index={0}
                              tabIndex={0}
                            >
                              <styled.div
                                active={false}
                                innerRef={[Function]}
                                onClick={[Function]}
                              >
                                <div
                                  className="c3"
                                  onClick={[Function]}
                                >
                                  Tab1
                                </div>
                              </styled.div>
                            </Tab>
                          </Component>
                        </sortableElement>
                      </DragTab>
                    </div>
                  </styled.div>
                </div>
              </styled.div>
            </div>
          </styled.div>
        </div>
      </TabList>
    </Component>
  </sortableList>
</DragTabList>
`;
